### PR TITLE
der, x509-cert: support for signed bigints, negative serial numbers

### LINKED ISSUE #821: x509-cert: parse failure on publicly known certificate
Ran into this in my continued adventures with Windows cert store parsing: the cert identified as [`28903a635b5280fae6774c0b6da7d6baa64af2e8`](http://www.download.windowsupdate.com/msdownload/update/v3/static/trustedr/en/28903a635b5280fae6774c0b6da7d6baa64af2e8.crt) fails with:

```
malformed ASN.1 DER value for INTEGER at DER byte 50
```

Here's the cert, in PEM form:

```pem
-----BEGIN CERTIFICATE-----
MIIFVjCCBD6gAwIBAgIQ7is969Qh3hSoYqwE893EATANBgkqhkiG9w0BAQUFADCB
8zELMAkGA1UEBhMCRVMxOzA5BgNVBAoTMkFnZW5jaWEgQ2F0YWxhbmEgZGUgQ2Vy
dGlmaWNhY2lvIChOSUYgUS0wODAxMTc2LUkpMSgwJgYDVQQLEx9TZXJ2ZWlzIFB1
YmxpY3MgZGUgQ2VydGlmaWNhY2lvMTUwMwYDVQQLEyxWZWdldSBodHRwczovL3d3
dy5jYXRjZXJ0Lm5ldC92ZXJhcnJlbCAoYykwMzE1MDMGA1UECxMsSmVyYXJxdWlh
IEVudGl0YXRzIGRlIENlcnRpZmljYWNpbyBDYXRhbGFuZXMxDzANBgNVBAMTBkVD
LUFDQzAeFw0wMzAxMDcyMzAwMDBaFw0zMTAxMDcyMjU5NTlaMIHzMQswCQYDVQQG
EwJFUzE7MDkGA1UEChMyQWdlbmNpYSBDYXRhbGFuYSBkZSBDZXJ0aWZpY2FjaW8g
KE5JRiBRLTA4MDExNzYtSSkxKDAmBgNVBAsTH1NlcnZlaXMgUHVibGljcyBkZSBD
ZXJ0aWZpY2FjaW8xNTAzBgNVBAsTLFZlZ2V1IGh0dHBzOi8vd3d3LmNhdGNlcnQu
bmV0L3ZlcmFycmVsIChjKTAzMTUwMwYDVQQLEyxKZXJhcnF1aWEgRW50aXRhdHMg
ZGUgQ2VydGlmaWNhY2lvIENhdGFsYW5lczEPMA0GA1UEAxMGRUMtQUNDMIIBIjAN
BgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAsyLHT+KXQpWIR4NA9h0X84NzJB5R
85iKw5K4/0CQBXCHYMkAqbWUZRkiFRfCQ2xmRJoNBD45b6VLeqpjt4pEndljkYRm
4CgPukLjbo73FCeTae6RDqNfDrHrZqJyTxIThmV6PttPB/SnCWDaOkKZx7J/sxaV
HMf5NLWUhdWZXqBIoH7nF2W4onW4HvPlQn2v7fOKSGRdghST2MDk/7NQcvJ29rNd
QlB50JQ+awwAvthrDk4q7D7SzIKiGGUzE3eeml0aE9jD2z3Il3rucO2n5nzbcc8t
lGLfbdb1OL4/pYUKGbio2Al1QnDE6u/LDsg0qBIimAy4E5S2S+zw0JDnJwIDAQAB
o4HjMIHgMB0GA1UdEQQWMBSBEmVjX2FjY0BjYXRjZXJ0Lm5ldDAPBgNVHRMBAf8E
BTADAQH/MA4GA1UdDwEB/wQEAwIBBjAdBgNVHQ4EFgQUoMOLRKo3pUW/l4Ba0fF4
opvpXY0wfwYDVR0gBHgwdjB0BgsrBgEEAfV4AQMBCjBlMCwGCCsGAQUFBwIBFiBo
dHRwczovL3d3dy5jYXRjZXJ0Lm5ldC92ZXJhcnJlbDA1BggrBgEFBQcCAjApGidW
ZWdldSBodHRwczovL3d3dy5jYXRjZXJ0Lm5ldC92ZXJhcnJlbCAwDQYJKoZIhvcN
AQEFBQADggEBAKBIW4IB9k1IuDlVNZyAelOZ1Vr/sXE7zDkJlF7W2u++AVtd0x7Y
/X1PzaBB4DSTv8vihpw3kpBWHNzrKQXlxJ7HNd+KDM3FIUPpqojlNcAZQmNaAl6k
SBg6hW/cnbw/nZzBh7h6YQjpdwt/cKt63dmXLGQehb+8dJahw3oS7AwaboMMPOhy
Rp/7SNVel+axofjk70YllJyJ22k4vuxcDlbHZVHlUIiIv0LVKz3l+bqeLrPK9HOS
Agu+TGbrIP65y7WZf+a2E/rKS03Z7lNGBjvGTq2TWoF+bCpLagVFjPIhpDGQh2xl
nJ2lYJU6Un/10asIbvPuW/mIPX64b24D5EI=
-----END CERTIFICATE-----
```

I can debug this on my own; just filing for visibility.

### PR DESCRIPTION
**Pull request by woodruffw**
~~WIP; needs unit tests. Just pushing up for visibility.~~

N.B.: My editor made some automatic reformatting decisions (I use `cargo +nightly fmt` by default), so let me know if you'd like me to revert them. They should be a superset of the existing `cargo fmt` style, however.

Fixes #821.

### CODE REVIEW

File: der/src/asn1/integer/int.rs [hunk 1]
```diff
@@ -34,14 +46,14 @@ where
...
-/// Get the encoded length for the given unsigned integer serialized as bytes.
+/// Get the encoded length for the given signed integer serialized as bytes.
 #[inline]
 pub(super) fn encoded_len(bytes: &[u8]) -> Result<Length> {
     Length::try_from(strip_leading_ones(bytes).len())
 }
 
 /// Strip the leading all-ones bytes from the given byte slice.
-fn strip_leading_ones(mut bytes: &[u8]) -> &[u8] {
+pub(crate) fn strip_leading_ones(mut bytes: &[u8]) -> &[u8] {
```
Comment (woodruffw) [inline der/src/asn1/integer/int.rs:56]: N.B.: I updated these comments; based on the context and behavior it looks like they were incorrectly copy-pasted before.

**Comment by woodruffw** [comment]
This should be good for a review; I've added some basic tests and confirmed that no other tests have broken with `SerialNumber` using an `Int` as its backing representation, but I'll also add the actual negative `SerialNumber` example in #821 as a smoke test.

**Review by tarcieri — APPROVED** [review]

**Comment by baloo** [comment]
I'm a bit late but, here is the relevant rfc piece I found about this https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.2,

anyway, this is in line with the rfc:
```
   Note: Non-conforming CAs may issue certificates with serial numbers
   that are negative or zero.  Certificate users SHOULD be prepared to
   gracefully handle such certificates.

**Comment by woodruffw** [comment]
Yep! I probably should have included a comment to that effect :slightly_smiling_face:

**Comment by tarcieri** [comment]
I can ask some relevant X.509 experts on recommendations for interpreting that SHOULD.

It would probably be best to have constructors only accept `Uint`, even if it can decode `Int`.

**Comment by woodruffw** [comment]
> It would probably be best to have constructors only accept `Uint`, even if it can decode `Int`.

Yeah, that makes sense to me. The API change I'm thinking about in https://github.com/RustCrypto/formats/issues/826#issuecomment-1367708814 would address this I think -- `Int` would become a sum type over signed/unsigned, and the `SerialNumber::new` ctor could then only accept the `Int(Unsigned(...))` variant.

**Pull request merged**